query/src/main/kotlin/com/linecorp/kotlinjdsl/query/creator/SubqueryCreatorImpl.kt

```diff
@@ -2,58 +2,23 @@ package com.linecorp.kotlinjdsl.query.creator
 
 import com.linecorp.kotlinjdsl.query.SubquerySpec
 import com.linecorp.kotlinjdsl.query.spec.Froms
-import javax.persistence.criteria.CriteriaBuilder
-import javax.persistence.criteria.CriteriaQuery
-import javax.persistence.criteria.CriteriaUpdate
-import javax.persistence.criteria.Subquery
+import javax.persistence.criteria.*
 
 class SubqueryCreatorImpl : SubqueryCreator {
     override fun <T> createQuery(
         spec: SubquerySpec<T>,
         froms: Froms,
-        criteriaQuery: CriteriaQuery<*>,
+        commonQuery: CommonAbstractCriteria,
```

Comment:
-- english
When I look the JPA interface structure and description of CommonAbstractCriteria, I am not sure if we can use this interface. It feels like an interface created by force to support Update and Delete.


-- korean
CommonAbstractCriteria를 보았을 때 우리가 이 인터페이스를 사용해도 될지 확신이 들지 않습니다. JPA 인터페이스 구조 및 설명을 읽어보면 Update 및 Delete를 지원하기 위해 억지로 만들어낸 인터페이스처럼 느껴지기 때문입니다.

Reply:
No problem at all. Because all the sub-implementations are already doing the implementation. The change in this specification is a situation that requires more changes, so there is no problem and the integration test is currently running, so it is correct to think that the function has been verified.

-- korean
전혀 문제가 없습니다. 이미 모든 하위 구현체들이 구현을 다 하고 있기 때문이죠. 이 스펙이 바뀐다는건 더 많은 변화가 필요한 상황일거라 문제가 없고 현재 integration test 도 돌고 있기 때문에 기능 검증도 된 상태라고 보는게 맞습니다.

Reply:
-- english
There is no problem with the function. But it's because I wonder if this interface was really created to use.

If JPA 3.X spec JPA 4.X spec determines that this interface is not good and removes it, I think our modification will be huge. We want to use the API's interface that supports sub-compatible.

-- korean
기능에 대해서는 문제가 없습니다. 하지만 이 인터페이스가 정말 사용하기 위해서 만들어진 인터페이스인지에 대해서 의문이 들기 때문입니다.

만약 JPA 3.X 스펙 JPA 4.X 스펙에서 이 인터페이스는 별로라고 판단하고 제거하면 우리의 여파도 엄청 커질 것이라 생각합니다. 이러한 내용 때문에 우리는 하위 호환성을 지원하는 API의 인터페이스를 사용하길 원하는데 주석으로는 하위 호환성을 지켜줄지 의문이 들기 때문입니다.

CommonAbstractCriteria
> It is not intended to be used directly in query construction.

Reply:
-- english
I think the problem is that an interface only for the extraction of common parts.

Dsl interface, for example, structure does not have a base. If we made BaseDsl, it would be much harder to use than now.

As such, we don't know how this interface will change in the future. There is a common method, so it's just grouped together.

If the role of this interface has not been defined like this, some people will wonder if it is okay to use it elsewhere.

Like the Common AbstractCriteria we're discussing below.

-- korean
공통 부분의 추출만을 위한 인터페이스가 되는 것이 문제라고 생각합니다.

Dsl 인터페이스를 예를 들면 지금 Dsl 인터페이스 구조는 Base가 없습니다. 만약 BaseDsl을 만들었다면 지금보다 훨씬 사용하기 힘들 것입니다. 

이처럼 이 인터페이스가 앞으로 어떻게 변경될지 모릅니다. 공통 메소드가 존재하니 하나로 묶었을 뿐입니다.

이처럼 이 인터페이스가 가진 역할이 정의 되지 않았다면 다른 곳에 사용할 때에도 정말 사용해도 괜찮은지 의문을 가지게 될 것입니다. 

마치 아래에서 의논하고 있는 CommonAbstractCriteria 처럼요

Reply:
-- english
I think it's better to get rid of such a possibility than to expect a very small possibility.

There is no disagreement about using CommonAbstractCriteria in the private method.

-- korean
극히 적다는 가능성을 우리가 기대하는 것보다 그런 가능성을 없애는 것이 더 좋다고 생각합니다.

private 메소드에서 CommonAbstractCriteria를 사용하는 것에는 이견이 없습니다.

Reply:
No If you look at the description of CommonAbstractCriteria, it says the following.

Note that criteria queries and criteria update and delete operations are typed differently. Criteria queries are typed according to the query result type. Update and delete operations are typed according to the target of the update or delete.

In other words, it is explained that the result of select is different depending on the return type, and update and delete are different depending on the from type.

This is because CriteriaUpdate and CriteriaDelete were created from 2.1 rather than forcing this interface.
Very few is just a different way of saying that there is really no such thing.

-- korean
아니요 CommonAbstractCriteria 의 설명을 보면 아래와 같은 내용이 있습니다.

Note that criteria queries and criteria update and delete operations are typed differently. Criteria queries are typed according to the query result type. Update and delete operations are typed according to the target of the update or delete.

즉, select 는 return type 에 따라 결과가 달라지고 update, delete 는 from type 에 따라 달라지기 때문이라고 설명이 적혀있습니다

이는 이 인터페이스가 억지로 만든게 아닌 CriteriaUpdate, CriteriaDelete 를 2.1 부터 만들었기 때문에 생긴것이고 
극히 적다는건 사실 없다는것을 다르게 쓴 말일 뿐입니다.

Reply:
-- english
This is because they can't combine Select, Update, and Delete into one query only by looking at the method and explanation, so I feel like I had no choice but to make it.

When I look at the interface description and the methods it has, I don't know what this interface is for.

-- korean
저는 메소드와 설명만 보았을 때 Select와 Update, Delete를 하나의 쿼리로 묶을 수 없으니 어쩔 수 없이 만들었다 라는 느낌으로 읽히기 때문입니다.

인터페이스 설명 및 인터페이스가 가진 메소드를 보았을 때 이 인터페이스는 무엇을 위해 존재하는지 모르겠습니다.

Reply:
Changing the upper interface of the interface is the same as the fact that the whole thing has to be overhauled.
If CommonAbstractCriteria changes, the rest of the changes will be huge. It's unlikely that that will change in the JPA specification, which I think will lead to much more discussion than we do. Rather, I don't know if there is another upper interface.

-- korean
인터페이스 상위가 변경 된다는건 전반적으로 다 뜯어 고쳐야 한다는 사실과 같습니다.
CommonAbstractCriteria 가 변경된다면 나머지 변경사항은 엄청날겁니다. 우리보다 훨씬 더 많은 토론이 이어질거라 생각되는 JPA 스펙에서 저게 변경될 가능성은 거의 없지 않을까 싶습니다. 오히려 상위 인터페이스가 또 생긴다면 모를까..

Reply:
If you look at the description of the interface, it seems that the parts that can be used in common are defined. Don't we call the code that has all the methods that can be used in common an interface?

-- korean
인터페이스의 설명을 보면 공통적으로 사용할 수 있는 부분을 정의한걸로 보이고요 우리는 모든 공통적으로 사용할 수 있는 메소드를 가진 코드를 인터페이스라고 하지 않나요?

Reply:
-- english
We concluded that the CommonAbstractCriteria interface does not bind CriteriaQuery, but rather binds Criteria.

All Criteria defined it as supporting subquery and decided to use it.

-- korean
CommonAbstractCriteria 인터페이스가 CriteriaQuery를 묶는 것이 아닌 Criteria를 묶는 것으로 결론을 내렸습니다.

모든 Criteria는 subquery를 지원하는 것으로 정의하고 criteria를 사용하기로 결정했습니다.